### PR TITLE
JAMES-3885 - Change username - should migrate succeed when mailboxes from old user already exist on the new user

### PR DESCRIPTION
**Pull request by vttranlina**
## Why

Script:
- Bob already has an inbox mailbox (no messages)
- Alice change her username to Bob by webadmin
- Get task status, the result is ABORTED
 
## How to fix it

Keep all old BOB mailbox data (messages, subscriptions...)

- Rename Bob X mailbox -> Bob X-migrating mailbox
- Rename Alice mailbox X -> Bob mailbox X
- Copy messages in Bob X-migrating mailbox to Bob mailbox X, and then delete Bob X-migrating mailbox

### CODE REVIEW

File: server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/MailboxUsernameChangeTaskStep.java [hunk 1]
```diff
@@ -76,11 +77,35 @@ public Publisher<Void> changeUsername(Username oldUsername, Username newUsername
 
     private Mono<Void> migrateMailbox(MailboxSession fromSession, MailboxSession toSession, org.apache.james.mailbox.model.MailboxMetaData mailbox) {
         MailboxPath renamedPath = mailbox.getPath().withUser(toSession.getUser());
-        return mailboxManager.renameMailboxReactive(mailbox.getPath(), renamedPath,
-            MailboxManager.RenameOption.RENAME_SUBSCRIPTIONS,
-            fromSession, toSession)
+        Mono<Void> renamePublisher = mailboxManager.renameMailboxReactive(mailbox.getPath(), renamedPath,
```
Comment (Arsnael) [inline server/container/mailbox-adapter/src/main/java/org/apache/james/adapter/mailbox/MailboxUsernameChangeTaskStep.java:80]: I feel that could be just extracted into a function directly?

File: server/container/mailbox-adapter/src/test/java/org/apache/james/adapter/mailbox/MailboxUsernameChangeTaskStepTest.java [hunk 2]
```diff
@@ -147,4 +151,51 @@ void shouldTransferSubscriptionToNewUser() throws Exception {
...
+        Mono.from(testee.changeUsername(ALICE, BOB)).block();
+
+        assertThat(mailboxManager.getMailbox(bobInbox, bobSession).getMessageCount(bobSession))
+            .isEqualTo(2);
+
+        assertThat(mailboxManager.list(bobSession)).containsOnly(bobInbox);
+    }
+
+    @Test
+    void shouldMigrateSubMailboxesWhenNewUserHasAlreadyMailbox() throws Exception {
```
Comment (Arsnael) [inline server/container/mailbox-adapter/src/test/java/org/apache/james/adapter/mailbox/MailboxUsernameChangeTaskStepTest.java:185]: If I understand correctly, Bob and Alice don't have the same mailboxes here though, so it's not really an already mailbox exists use case?

**Review by Arsnael — COMMENTED** [review]
Also the commit description is not accurate:

```
JAMES-3885 - Change username - should migrate succeed when new user already exists
```
To do the migration, the new user needs to exist, or it fails. Here it's more if mailboxes from old user already exist on the new user :)

**Review by chibenwa — APPROVED** [review]

File: server/container/mailbox-adapter/src/test/java/org/apache/james/adapter/mailbox/MailboxUsernameChangeTaskStepTest.java (hunk 2 above)
Comment (vttranlina) [reply server/container/mailbox-adapter/src/test/java/org/apache/james/adapter/mailbox/MailboxUsernameChangeTaskStepTest.java:185]: > Bob and Alice don't have the same mailboxes here though,

No, it abort when Bob and Alice have the same mailbox name

**Comment by quantranhong1999** [comment]
```
Rename Bob X mailbox -> Bob X-migrating mailbox
Rename Alice mailbox X -> Bob mailbox X
Copy messages in Bob X-migrating mailbox to Bob mailbox X, and then delete Bob X-migrating mailbox
```

Rethinking about this, could we just move directly messages from `Alice mailbox X` to `Bob mailbox X` (without a temporary mailbox)?

**Comment by vttranlina** [comment]
> Rethinking about this, could we just move directly messages from `Alice mailbox X` to `Bob mailbox X` (without a temporary mailbox)?

The rename method not only moves messages
It also subscription, acl, sub mailbox, delegatee...vvv
I believe we have more than one way to do this, but the renameMethod looks complex, so let's make it less code

File: server/container/mailbox-adapter/src/test/java/org/apache/james/adapter/mailbox/MailboxUsernameChangeTaskStepTest.java (hunk 2 above)
Comment (Arsnael) [reply server/container/mailbox-adapter/src/test/java/org/apache/james/adapter/mailbox/MailboxUsernameChangeTaskStepTest.java:185]: I thought your change was supposed to be able to do the migration when mailboxes have the same name? I don't get your comment sorry

Comment (vttranlina) [reply server/container/mailbox-adapter/src/test/java/org/apache/james/adapter/mailbox/MailboxUsernameChangeTaskStepTest.java:185]: Ah, I just re-read your comment, 
This test case for ensuring when user has mailbox A, 
Mailbox A won't be changed when migrate mailbox B

It look like not clear sense, maybe it is unnecessary?

Comment (Arsnael) [reply server/container/mailbox-adapter/src/test/java/org/apache/james/adapter/mailbox/MailboxUsernameChangeTaskStepTest.java:185]: Ok I got it :)

How about renaming the test: `shouldMigrateMailboxesWhenNewUserHasAlreadyOherMailboxes` ?

**Review by Arsnael — APPROVED** [review]

**Review by quantranhong1999 — APPROVED** [review]

**Pull request merged**